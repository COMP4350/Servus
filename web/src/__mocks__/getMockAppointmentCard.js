function getMockAppointmentCard() {
    return {
        buyer: 'Risto',
        provider: 'Arveend',
        service_id: '1001',
        created_at: new Date(100),
        booked_time: new Date(500),
    }
}

export const mockAppointmentCard = getMockAppointmentCard();