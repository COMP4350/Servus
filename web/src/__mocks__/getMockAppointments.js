function getMockAppointments() {
    return [
        {
            buyer: 'Risto',
            provider: 'Arveend',
            service_id: '1001',
            created_at: new Date(100),
            booked_time: new Date(500),
        },
        {
            buyer: 'Arveend',
            provider: 'Risto',
            service_id: '1002',
            created_at: new Date(700),
            booked_time: new Date(1200),
        },
        {
            buyer: 'Caden',
            provider: 'Andy',
            service_id: '1003',
            created_at: new Date(1400),
            booked_time: new Date(1900),
        },
    ]
}

function getMockAppointmentsCookie() {
    return {
        username: 'testusername',
    }
}

export const mockAppointments = getMockAppointments();
export const mockAppointmentsCookie = getMockAppointmentsCookie();