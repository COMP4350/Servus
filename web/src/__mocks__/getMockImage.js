function getMockImage() {
    return {
        ownerUsername: 'riki',
        uploadDate: new Date(2012, 1, 1),
        imageName: 'example_image.jpeg',
        imageData: 'the/link/to/data',
        profilePicture: false,
    };
}

export const mockImage = getMockImage();
