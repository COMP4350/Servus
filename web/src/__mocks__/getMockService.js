function getMockService() {
    return {
        _id: '1001',
        name: "Risto's Painting",
        provider: 'Risto',
        description: 'Risto paints you',
        cost: 100,
        duration: '0100',
        availability: [{ weekday: 1, start_time: '0800', end_time: '1800' }],
        ratings: [{ username: 'arvind', rating: 5 }],
    };
}

export const mockService = getMockService();
