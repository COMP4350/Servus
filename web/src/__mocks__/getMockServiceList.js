function getMockServiceList() {
    return [
        {
            name: "Risto's Painting",
            provider: 'Risto',
            description: 'Risto paints you',
            cost: 100,
            duration: 60,
        },
        {
            name: "Reesto's Painting",
            provider: 'Reesto',
            description: 'Reesto paints yee',
            cost: 120,
            duration: 80,
        },
        {
            name: "Rasta's Painting",
            provider: 'Rasta',
            description: 'Rasta paints yaa',
            cost: 150,
            duration: 90,
        },
    ];
}

export const mockServiceList = getMockServiceList();
