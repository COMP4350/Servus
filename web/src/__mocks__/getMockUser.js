function getMockUser() {
    return {
        username: 'testuser',
        firstName: 'test',
        lastName: 'user',
        password: '123',
        bio: 'im a test user',
    };
}

export const mockUser = getMockUser();
