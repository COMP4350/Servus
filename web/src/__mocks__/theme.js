import { createMuiTheme } from '@material-ui/core/';

export default createMuiTheme({
    palette: {
        primary: {
            main: '#647AA3',
        },
        secondary: {
            main: '#EC5732',
        },
        default: {
            main: '#FFF2EB',
        },
    },
    background: {
        dark: '#151515',
        main: '#272727',
    },
    typographs: {
        fontFamily: ['Roboto'],
    },
});
