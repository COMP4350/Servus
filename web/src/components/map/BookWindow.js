import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { makeStyles, createMuiTheme } from '@material-ui/core/styles';
import {
    ThemeProvider,
    Button,
    Select,
    InputLabel,
    MenuItem,
    FormControl,
    Typography,
} from '@material-ui/core';
import moment from 'moment';
import MomentUtils from '@date-io/moment';
import { MuiPickersUtilsProvider, DatePicker } from '@material-ui/pickers';
import { useCookies } from 'react-cookie';

const theme = createMuiTheme({
    palette: {
        type: 'dark',
        primary: {
            main: '#647AA3',
        },
        secondary: {
            main: '#EC5732',
        },
        default: {
            main: '#FFF2EB',
        },
    },
    background: {
        dark: '#151515',
        main: '#272727',
    },
    typography: {
        fontFamily: ['Roboto'],
    },
});

const useStyles = makeStyles(() => ({
    window: {
        margin: '50',
        display: 'flex',
        alignItems: 'center',
        'flex-direction': 'column',
        width: '100%',
        height: '100%',
        background: theme.background.main,
    },
    innerwindow: {
        display: 'flex',
        height: '100%',
        'flex-direction': 'column',
        justifyContent: 'space-evenly',
    },
    container: {
        display: 'flex',
        alignItems: 'center',
        'flex-direction': 'column',
    },
    textField: {
        marginLeft: theme.spacing(1),
        marginRight: theme.spacing(1),
        width: '100%',
    },
    formControl: {
        marginLeft: theme.spacing(1),
        marginRight: theme.spacing(1),
        width: '100%',
        height: 40,
    },
    userIcon: {
        height: '96px',
        width: '96px',
    },
    title: {
        marginTop: '15px',
        fontSize: '2.5em',
        color: 'white',
    },
    divider: {
        height: '100%',
        width: '3',
        'border-color': theme.background.main,
    },
    close: {
        position: 'absolute',
        top: 0,
        right: 10,
        color: 'white',
        fontSize: '2em',
    },
    submit: {},
}));

const BookWindow = props => {
    // STATES
    const [date, setDate] = useState();
    const [timeState, setTime] = useState({
        time: '',
    });
    const [timepicker, setTimepicker] = useState();
    const classes = useStyles();
    const [valid, setValid] = useState(false);
    const [errors, setErrors] = useState({});

    const [cookies] = useCookies();


    let duration_hours = props.service.duration.slice(0, 2);
    let duration_minutes = props.service.duration.slice(2);

    const displayTimePicker = async () => {
        const removeItemOnce = (arr, value) => {
            let index = arr.indexOf(value);
            if (index > -1) {
                arr.splice(index, 1);
            }
            return arr;
        };
        
        let time_array = [];
        const addTimeSlices = () => {
            for (let avail in props.service.availability) {
                if (props.service.availability[avail].weekday == date.day()) {
                    let start_time = moment(
                        props.service.availability[avail].start_time,
                        'HH:mm'
                    );
                    let end_time = moment(
                        props.service.availability[avail].end_time,
                        'HH:mm'
                    );
                    let apt_time = moment(start_time)
                        .add(duration_hours, 'h')
                        .add(duration_minutes, 'm');
                    let curr_time = start_time;
                    while (apt_time <= end_time) {
                        time_array.push(moment(curr_time).format('HH:mm'));
                        curr_time.add(30, 'm');
                        apt_time.add(30, 'm');
                    }
                }
            }
        };
        const removeConflictingTimes = async () => {
            let res = await axios.get(
                `appointment/service/${props.service._id}`
            );

            let available_times = [];
            for (let appt in res.data.result) {
                let apt_date = moment(res.data.result[appt].booked_time);
                let apt_end_date = moment(apt_date)
                    .add(duration_hours, 'h')
                    .add(duration_minutes, 'm');
                if (apt_date.isSame(date, 'day')) {
                    for (let i in time_array) {
                        let extracted_date = date.format('YYYY-MM-DD');
                        let extracted_time = moment(
                            time_array[i],
                            'HH:mm'
                        ).format('HH:mm');
                        let booking_date = moment(
                            `${extracted_date} ${extracted_time}`,
                            'YYYY-MM-DD HH:mm'
                        );
                        let booking_date_end = moment(booking_date)
                            .add(duration_hours, 'h')
                            .add(duration_minutes, 'm');

                        if (
                            (booking_date >= apt_date &&
                                booking_date <= apt_end_date) ||
                            (booking_date_end >= apt_date &&
                                booking_date_end <= apt_end_date)
                        ) {
                            available_times.push(time_array[i]);
                        }
                    }
                }
            }
            for (let i in available_times) {
                removeItemOnce(time_array, available_times[i]);
            }
        };

        addTimeSlices();
        await removeConflictingTimes();

        setTimepicker(
            <FormControl className={classes.formControl}>
                <InputLabel id="label" className={classes.inputLabel}>
                    Appointment Time
                </InputLabel>
                <Select
                    labelId="label"
                    value={timeState.time}
                    onChange={x => handleChangeTime(x)}>
                    {time_array?.map((x, i) => {
                        return (
                            <MenuItem key={i} value={x}>
                                {x}
                            </MenuItem>
                        );
                    })}
                </Select>
            </FormControl>
        );
    };

    const bookAppointment = () => {
        if (valid) {
            let extracted_date = date.format('YYYY-MM-DD');
            let extracted_time = moment(timeState.time, 'HH:mm').format(
                'HH:mm:ss'
            );

            let booking_date = moment(
                `${extracted_date} ${extracted_time}`,
                'YYYY-MM-DD HH:mm:ss'
            ).toDate();
            axios
                .post(
                    `/appointment/${cookies.username}`,
                    {
                        service_id: props.service._id,
                        provider: props.service.provider,
                        booked_time: booking_date,
                    },
                    {
                        withCredentials: true,
                    }
                )
                .then(res => {
                    if (!res.data.errors) {
                        alert(
                            `Successfully booked Service with ${props.service.provider}`
                        );
                    }
                })
                .catch(err => {
                    throw err;
                });
            props.toggle();
            setValid(false);
        }
    };

    useEffect(() => {
        if (cookies.username) bookAppointment();
        if (date) displayTimePicker();
    }, [valid, date, timeState]);

    const handleChangeTime = x => {
        setTime({ time: x.target.value });
    };

    const shouldDisableDay = date => {
        for (let avail in props.service.availability) {
            if (props.service.availability[avail].weekday == date.day())
                return false;
        }
        return true;
    };
    const validate = () => {
        let errors = {};
        if (!date || !timeState.time) {
            errors.time = 'Date & Time are required';
        }
        setErrors(errors);
        setValid(Object.getOwnPropertyNames(errors).length == 0);
    };

    const handleClick = () => {
        props.toggle();
    };

    let apptDuration = `Appointment Duration:`;
    if(duration_hours === "00")
        apptDuration = `${apptDuration} ${duration_minutes} mins`;
    else
        apptDuration = `${apptDuration} ${duration_hours}:${duration_minutes} hrs`;

    console.log()
    return (
        <ThemeProvider theme={theme}>
            <div className={classes.window}>
                <Typography variant="h1" className={classes.title}>
                    {props.service.name}
                </Typography>
                <Typography color="textSecondary">
                    { apptDuration }
                </Typography>
                <div className={classes.innerwindow}>
                    <form className={classes.container} noValidate>
                        <MuiPickersUtilsProvider utils={MomentUtils}>
                            <DatePicker
                                label={'Appointment Date'}
                                value={date}
                                initialFocusedDate={Date.now()}
                                onAccept={setDate}
                                onChange={setDate}
                                name="date"
                                autoOk={true}
                                disablePast={true}
                                errors={errors.time}
                                shouldDisableDate={shouldDisableDay}
                                className={classes.textField}
                            />
                            {timepicker}
                        </MuiPickersUtilsProvider>
                    </form>
                    <Button
                        variant="contained"
                        color="primary"
                        className={classes.submit}
                        onClick={validate}>
                        CONFIRM
                    </Button>
                </div>
                <span className={classes.close} onClick={handleClick}>
                    &times;{' '}
                </span>
            </div>
        </ThemeProvider>
    );
};

export default BookWindow;
