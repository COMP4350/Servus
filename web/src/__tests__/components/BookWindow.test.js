import React from 'react';
import { render } from '@testing-library/react';
import BookWindow from '../../components/map/BookWindow';
import { mockService } from '../../__mocks__/getMockService';

const blankComponent = () => {
    return <div></div>;
};
jest.mock('@material-ui/pickers', () => {
    return {
        __esModule: true,
        DatePicker: () => {
            return blankComponent();
        },
        MuiPickersUtilsProvider: () => {
            return blankComponent();
        },
    };
});

describe('BookWindow', () => {
    it('renders correctly', () => {
        const { asFragment } = render(<BookWindow service={mockService} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
