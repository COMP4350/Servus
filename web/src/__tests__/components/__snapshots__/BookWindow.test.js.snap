// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookWindow renders correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-window-1"
  >
    <h1
      class="MuiTypography-root makeStyles-title-7 MuiTypography-h1"
    >
      Risto's Painting
    </h1>
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Appointment Duration: 01:00 hrs
    </p>
    <div
      class="makeStyles-innerwindow-2"
    >
      <form
        class="makeStyles-container-3"
        novalidate=""
      >
        <div />
      </form>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-10 MuiButton-containedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          CONFIRM
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <span
      class="makeStyles-close-9"
    >
      Ã— 
    </span>
  </div>
</DocumentFragment>
`;
